### PR TITLE
LPS-158484 Add Test Autom. ExposeErcFromAssetLibraryForBlog#canCreateCommentWithCustomExternalReferenceCode

### PR DESCRIPTION
**Pull request by mateusralv**
**Background:** 
Create automated test ExposeErcFromAssetLibraryForBlog#canCreateCommentWithCustomExternalReferenceCode and ExposeErcFromAssetLibraryForBlog#canNotCreateTwoCommentsWithCustomExternalReferenceCode

**Test Plan:**
**Given** blog post is created
**When** with POST request I create a comment with a custom external reference code
**Then** comment is being created
**And Then** I can see the custom external reference code in the body response

**Given** blog post is created
**And Given** comment with a custom external reference code is created with a POST request
**When** with POST request I create a comment with an already existing custom external reference code
**Then** I receive an error code response
**And Then** another comment with the same external reference code is not being created

**JIRA Link:**
https://issues.liferay.com/browse/LPS-159571
https://issues.liferay.com/browse/LPS-158484

### CODE REVIEW

**Comment by mateusralv** [comment]
ci:test:headless-frontend-functional:nocompile

**Comment by mateusralv** [comment]
ci:test:headless-frontend-functional:nocompile

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlog.macro [hunk 1]
```diff
@@ -30,4 +31,64 @@ definition {
...
+
+	macro assertCommentTextWithCorrectValue {
+		var actualTextComments = JSONUtil.getWithJSONPath("${blogsPostingId}", "$..['items'][?(@.creator)].text");
+
+		TestUtils.assertEquals(
+			actual = "${actualTextComments}",
+			expected = "${actualTextComments}");
+	}
+
+	macro assertProperNumberOfItems {
```
Comment (magjed4289) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlog.macro:50]: we already have the macro BlogPostingAPI.getNumberOfBlogPostingsCreated(); which returns a number of blogs created. You can then use it directly in the test step, example:

```
task ("And Then the 2 blog postings are visible as created") {
			var actualNumberOfBlogPostings = BlogPostingAPI.getNumberOfBlogPostingsCreated();

			TestUtils.assertEquals(
				actual = "${actualNumberOfBlogPostings}",
				expected = "2");
		}


```

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlog.macro [hunk 2]
```diff
@@ -30,4 +31,64 @@ definition {
 		return "${entryId}";
 	}
 
+	macro assertCommentERCWithCorrectValue {
```
Comment (magjed4289) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlog.macro:34]: you can use `BlogPostingAPI._getFieldValueOfExistingBlogPosting()` to get the externalReferenceCode, example:

var response (vía BlogPostingAPI.getBlogPostings(); or passing the response from another query)

```
var existingExternalReferenceCode = BlogPostingAPI._getFieldValueOfExistingBlogPosting(
					fieldName = "externalReferenceCode",
					response = "${response}");
```

**Review by magjed4289 — CHANGES_REQUESTED** [review]
Please take a look at BlogPostingAPI.macro and think about what values you could reuse without creating a new one.

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlogAPI.macro [hunk 3]
```diff
@@ -87,4 +92,67 @@ definition {
 		return "${entryId}";
 	}
 
+	macro _getAllBlogsPostingId {
```
Comment (magjed4289) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlogAPI.macro:95]: maybe `BlogPostingAPI.getBlogPostingsIdList()` could be helpful here. With this method `_getAllBlogsPostingId` what I see you receive is the whole response with blog postings body, not only the ID.

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlog.macro [hunk 4]
```diff
@@ -30,4 +31,64 @@ definition {
...
+		var actualCommentERCValue = JSONUtil.getWithJSONPath("${response}", "$..['items'][?(@.creator)].externalReferenceCode");
+
+		TestUtils.assertEquals(
+			actual = "${actualCommentERCValue}",
+			expected = "${expectedCommentERCValue}");
+	}
+
+	macro assertCommentTextWithCorrectValue {
```
Comment (magjed4289) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlog.macro:42]: same here, try to create a generic method for all the values of the body

**Review by magjed4289 — CHANGES_REQUESTED** [review]
Please review BlogPostingAPI file and if any new macro for BlogPosting needed, add it there. For everything regarding the Comment, please rename the file to CommentAPI.macro and Comment.macro (if not yet existing). This way we can have the files better organised! :)

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlog.macro (hunk 2 above)
Comment (mateusralv) [reply portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlog.macro]: Great idea !!! Thank u Meg.

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlog.macro (hunk 4 above)
Comment (mateusralv) [reply portal-web/test/functional/com/liferay/portalweb/macros/headless/blogs/HeadlessBlog.macro]: Done!

**Comment by mateusralv** [comment]
ci:test:headless-frontend-functional:nocompile

**Comment by magjed4289** [comment]
Just started reviewing :)

**Review by magjed4289 — APPROVED** [review]
Hey Mateus, I did a little moving of the macros to the corresponding files and also used the opportunity to refactor one of the methods I use in batchEngine testing. 
Take a look and if you have any questions, let me know!

**Comment by magjed4289** [comment]
ci:test:headless-frontend-functional:nocompile

**Comment by yanan** [comment]
ci:test:headless-frontend-functional:nocompile

**Comment by mateusralv** [comment]
ci:forward

**Comment by mateusralv** [comment]
ci:test:sf

**Comment by yanan** [comment]
ci:test:sf

**Comment by yanan** [comment]
ci:test:headless-frontend-functional:nocompile

**Comment by yanan** [comment]
ci:test:relevant

**Comment by yanan** [comment]
Hi @mateusralv , the previous sf failure is because there is one commit without ticket number, I've force pushed a fix. And the current headless-frontend-functional tests failures are unrelated to the current pr, most of the failure are because of [LPS-160768](https://issues.liferay.com/browse/LPS-160768). I'm going to forward this one after get Meg's review last week. cc @magjed4289

**Comment by yanan** [comment]
ci:forward

**Pull request closed without merging**